Added signing-in, default avatars, support for guest accounts, quoting and replying to messages

arnav-t opened the pull request:
## Added support for signing-in
### Via sign-in popup
![image](https://user-images.githubusercontent.com/34343421/85805900-3530dd80-b76b-11ea-86d3-c9b4ea18f2e1.png)
### Via `postMessage` interface
```js
iframe.contentWindow.postMessage({
    cmd: 'login',
    args: {
        user: 'username',
        passwd: 'password'
    }
}, origin);
```
*Arguments:*

`args` (object):   
* `username` (string) - Username
* `passwd` (string) - Password   

*Response:*
```js
{
    status: 'success', 
    message: 'Attempting sign in...'
}
```
## Added support for default avatars
Using the same design as Riot. However, I am just using flexbox and text instead of drawing on an HTML5 canvas. The HSL color is generated by hashing the user ID.
![image](https://user-images.githubusercontent.com/34343421/85806272-38789900-b76c-11ea-8d81-2130eb08122b.png)
## Quoting and replying
Added on-hover buttons and reply popup above message composer similar to Riot web. Quoting prepends the message (quoted) to the message composer similar to Riot web and replying would also work similar to Riot web. It will be fully functional once markdown parsing is added.  
![image](https://user-images.githubusercontent.com/34343421/85806621-33681980-b76d-11ea-8c1d-363d6cceda32.png)
![image](https://user-images.githubusercontent.com/34343421/85806928-09632700-b76e-11ea-9f5c-df8101e4267c.png)
## Guest access
The client now supports guest access. If no access token or user ID is provided in the configuration file, the client attempts to register a guest account on the home server. Room contents are viewed using `peekInRoom`. 
If the guest attempts to send a message, only then `joinRoom` is called to avoid spam (each page load would lead to a guest joining the room).
![image](https://user-images.githubusercontent.com/34343421/85807341-41b73500-b76f-11ea-92b0-30ddafb42a64.png)

arnav-t commented:
There is still a small bug with guest accounts as I described here: https://matrix.to/#/!JAXbbAQpdHGaamFkTO:t2l.io/$VhjzSU6EnKHZjYKS6ZdlL3AgtY6FSmrt0eQePW7JkiI?via=matrix.org&via=t2l.io&via=bpulse.org

benparsons commented:
![Screenshot 2020-06-28 at 12 57 24](https://user-images.githubusercontent.com/1197991/85946856-30bd1e00-b93f-11ea-8626-81858504c110.png)

Interesting bug - the message is duplicated in the client it was sent from. Refreshing cleared the problem, and I could not reproduce.

benparsons commented:
Re replies, is it know that the replied message is not displayed? Is this what you refer to by

> It will be fully functional once markdown parsing is added.

?

benparsons commented on components/reply-popup.jsx:
```diff
@@ -0,0 +1,79 @@
+        this.unsetReply = this.unsetReply.bind(this);
+    }
+
+    unsetReply() {
```
This needs to be called after message send, otherwise the reply box stays.

arnav-t replied:
Oh right, I'll fix that.

arnav-t commented:
@benparsons 
What I mean by that is that I only need to construct the formatted `content` object before sending the message. Since both quoting and replying send `formatted_body`, It would probably be best to have the markdown parser generate that.
For example, in the case of quoted replies a message like this `'> quote\n\nmy message'` should have a `content` object like this:
```js
{				
    body: "> quote\n\nmy message",
    format: "org.matrix.custom.html",
    formatted_body: "<blockquote>quote</blockquote>my message",
    msgtype: "m.text"
}
```
The `formatted_body` can be very easily generated using a markdown parser that will convert the `body` to `formatted_body`.

In the case of replies, `content` should look like this:
```js
{				
    body: "> <@ID_of_OP:server.com> ORIGINAL_MESSAGE\n\nREPLY_MESSAGE",
    format: "org.matrix.custom.html",
    formatted_body: "<mx-reply><blockquote><a href=\"https://matrix.to/#/!KgwjztkMFukvThzynq:asra.gr/$7iS956hLeVXyxbyQEiQA1zeuZN9p8ksvdEiC9y0Uyac?via=asra.gr\">In reply to</a> <a href=\"https://matrix.to/#/@ID_of_OP:server.com\">@ID_of_OP:server.com</a><br>ORIGINAL_MESSAGE</blockquote></mx-reply>REPLY_MESSAGE",
    m.relates_to: {
        m.in_reply_to: {
            event_id: "EVENT_ID_of_ORIGINAL_MESSAGE"
        }
    },
    msgtype: "m.text"
}
```
The `body` and `formatted_body` are slightly different here but it also might be better handled by the parser.

arnav-t commented on components/message-composer.jsx:
```diff
@@ -52,6 +54,8 @@ export default class MessageComposer extends PureComponent {
             busy: true
         });
+        this.props.client.stopPeeking();
+        await this.props.client.joinRoom(this.props.roomId, {syncRoom: true});
```
> ![Screenshot 2020-06-28 at 12 57 24](https://user-images.githubusercontent.com/1197991/85946856-30bd1e00-b93f-11ea-8626-81858504c110.png)
> 
> Interesting bug - the message is duplicated in the client it was sent from. Refreshing cleared the problem, and I could not reproduce.

@benparsons 
I'm not completely sure why this happens. Since the client is running in guest mode here, whenever someone attempts to leave a comment I call `stopPeeking` and then `joinRoom` just to make sure they're actually a member of the room before sending the event. The issue only presents itself when they actually weren't a member before and end up joining the room. It's probably got something to do with this. 
Either the wrong name appears on the first message (the user ID instead of the name, but is correct for subsequent messages) or the messages appear to be duplicated. These only show up in the client the message was sent from and seem to disappear on refreshing.

arnav-t commented:
@benparsons @turt2live 
I have added a call to `unsetReply` after sending a message as suggested by @benparsons 
I have removed the peeking code and changed it so the guest just joins the room straight away as suggested by @turt2live which means the bug has been fixed.
I have updated the demo too with these changes.

arnav-t commented:
Merged! :tada:
